### PR TITLE
Ncsdk 16156 sensor stub

### PR DESCRIPTION
**Pull request by rakons**
Cleanup caf_sensor_manager sample to simplify sensor choice.
Implement sensor stub.

### CODE REVIEW

File: samples/caf_sensor_manager/app.overlay [hunk 1]
```diff
@@ -9,6 +9,7 @@
 		compatible = "nordic,sensor-sim";
 		label = "SENSOR_SIM";
 		acc-signal = "wave";
+		status = "okay";
```
Comment (mbolivar-nordic) [inline samples/caf_sensor_manager/app.overlay:12]: Why is this relevant to this commit? The default status is `okay` if not defined, so this is redundant and does not seem related to the commit's purpose as I understand it.

File: dts/bindings/sensor/nordic,sensor-stub.yaml [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# Copyright (c) 2022 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+
+description: Simple sensor simulator
```
Comment (mbolivar-nordic) [inline dts/bindings/sensor/nordic,sensor-stub.yaml:5]: I have no objections to this approach going into NCS, but I would like you to be aware of the ongoing upstream work on "emulators" as well in case they are relevant to you.

The difference between emulators and stubs as I understand it is that you can write an emulator driver and a real device driver for the same devicetree compatible, and then choose at build time whether to use the emulator or device driver. This can make DTS management simpler if you have an emulator defined for some sort of "default" sensor device, if one is applicable. I don't know your use cases, though, so I cannot comment on whether this approach is better for you.

File: samples/caf_sensor_manager/app.overlay (hunk 1 above)
Comment (rakons) [reply samples/caf_sensor_manager/app.overlay:12]: This was relevant because the way the sensor_sim is enabled now uses the default Kconfig setting that is delivered by devicetree check for this status. This however may not be relevant as I am having issues with this way of configuration in more complex architectures.

File: dts/bindings/sensor/nordic,sensor-stub.yaml (hunk 2 above)
Comment (rakons) [reply dts/bindings/sensor/nordic,sensor-stub.yaml:5]: Sensor stub here is kind of minimal implementation. I try to implement sensor simulation with the minimal amount of code and put all the waveform generation to the application. This is because we wish to get rid of float point operation in some implementations and squeeze the FLASH usage.

File: samples/caf_sensor_manager/app.overlay (hunk 1 above)
Comment (rakons) [reply samples/caf_sensor_manager/app.overlay:19]: I had changed the implementation for the configuration be only in Kconfig. The DTS now defines kind of "maximal" configuration. Please recheck.

File: samples/caf_sensor_manager/sample.yaml [hunk 3]
```diff
@@ -18,6 +18,23 @@ tests:
         - "main state:READY"
         - "Send sensor buffer desc address:"
         - "sensor_data_aggregator_release_buffer_event"
+  sample.caf_sensor_manager.sensor_stub.correctness_test:
+    build_only: false
+    harness: console
+    platform_allow: nrf52840dk_nrf52840 nrf5340dk_nrf5340_cpuapp qemu_cortex_m3
```
Comment (emob-nordic) [inline samples/caf_sensor_manager/sample.yaml:24]: How is this gonna work with quemu platform if it has one core?

File: samples/caf_sensor_manager/src/modules/sensor_stub_gen.c [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+		*sample = gen_data[SENSOR_STUB_CH_Z].val;
+		return 0;
+	default:
+		return -ENOTSUP;
+	}
+}
+
+int sensor_stub_gen_init(const struct device *dev)
+{
+	LOG_INF("READY");
```
Comment (emob-nordic) [inline samples/caf_sensor_manager/src/modules/sensor_stub_gen.c:170]: Nothing happens here so maybe `LOG_DBG`?

File: samples/caf_sensor_manager/src/modules/sensor_stub_gen.c [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+	}
+
+	fp_2_sensor_val(&ch_data->val, val);
+}
+
+int sensor_stub_gen_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	switch (chan) {
+	case SENSOR_CHAN_ACCEL_XYZ:
+	case SENSOR_CHAN_ALL:
```
Comment (emob-nordic) [inline samples/caf_sensor_manager/src/modules/sensor_stub_gen.c:110]: Fetch supports `SENSOR_CHAN_ALL` but get does not, why is that?

File: samples/caf_sensor_manager/src/modules/sensor_stub_gen.c [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+#include <zephyr/kernel.h>
+#include <drivers/sensor_stub.h>
+
+#define MODULE sensor_stub_gen
+#include <caf/events/module_state_event.h>
+
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_SENSOR_STUB_GEN_LOG_LEVEL);
```
Comment (emob-nordic) [inline samples/caf_sensor_manager/src/modules/sensor_stub_gen.c:15]: Looking into logged messages im wondering if we need this at all? What do you think?

**Review by emob-nordic — CHANGES_REQUESTED** [review]

File: samples/caf_sensor_manager/src/modules/sensor_stub_gen.c [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+}
+
+static inline void fp_2_sensor_val(struct sensor_value *out, int64_t val)
+{
+	out->val1 = (int32_t)(val / SENSOR_FRACTIONAL_MAX);
+	out->val2 = (int32_t)(val % SENSOR_FRACTIONAL_MAX);
+}
+
+/* Generate changes on selected channel with given parameters */
+static inline void sensor_stub_gen_chan(struct sensor_stub_gen_ch_data *ch_data,
```
Comment (zycz) [inline samples/caf_sensor_manager/src/modules/sensor_stub_gen.c:86]: Maybe sensor_stub_gen_changes - current name can be easily confused chan -> channel

File: samples/caf_sensor_manager/sample.yaml (hunk 3 above)
Comment (zycz) [reply samples/caf_sensor_manager/sample.yaml:24]: Why one platform is a problem here? It works on 52840 which also has one core.

File: samples/caf_sensor_manager/remote/sensor_stub_overlay.conf [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+# Copyright (c) 2022 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+################################################################################
+# Application configuration overlay to select Sensor Stub
+
+# Select Sensor Stub
+CONFIG_SENSOR_SIM=n
+CONFIG_SENSOR_STUB=y
```
Comment (zycz) [inline samples/caf_sensor_manager/remote/sensor_stub_overlay.conf:11]: Do we need this file both for app and remote core? Cannot we have it once and used twice?

File: samples/caf_sensor_manager/app.overlay [hunk 9]
```diff
@@ -5,6 +5,13 @@
  */
 
  / {
+	sensor_stub: sensor_stub {
+		compatible = "nordic,sensor-stub";
+		label = "SENSOR_STUB";
+		generator = "sensor_stub_gen";
+		status = "okay";
```
Comment (zycz) [inline samples/caf_sensor_manager/app.overlay:12]: Should both simulators have status okay? Wouldn't it be confusing for a user?

File: samples/caf_sensor_manager/prj.conf [hunk 10]
```diff
@@ -18,6 +18,10 @@ CONFIG_APP_EVENT_MANAGER=y
 CONFIG_REBOOT=y
 CONFIG_HEAP_MEM_POOL_SIZE=4096
 
+# Using simulated sensor (the DK does not have built-in sensor)
+CONFIG_SENSOR_SIM=y
+CONFIG_SENSOR_STUB=n
```
Comment (zycz) [inline samples/caf_sensor_manager/prj.conf:23]: You are changing CONFIG_SENSOR_SIM in two commits it can be confusing.

File: samples/caf_sensor_manager/boards/nrf5340dk_nrf5340_cpuapp.conf [hunk 11]
```diff
@@ -18,4 +18,3 @@ CONFIG_CAF_SENSOR_DATA_AGGREGATOR=n
 CONFIG_CAF_SENSOR_DATA_AGGREGATOR_EVENTS=y
 
 CONFIG_SENSOR=n
-CONFIG_SENSOR_SIM=n
```
Comment (zycz) [inline samples/caf_sensor_manager/boards/nrf5340dk_nrf5340_cpuapp.conf:21]: Shouldn't we leave this line? In prj.conf we set CONFIG_SENSOR_SIM=y and in multicore variant of sample we don't need it.

**Review by zycz — CHANGES_REQUESTED** [review]

File: samples/caf_sensor_manager/src/modules/sensor_stub_gen.c (hunk 5 above)
Comment (rakons) [reply samples/caf_sensor_manager/src/modules/sensor_stub_gen.c:110]: Because you cannot get all the channels at once. This way it also works in real drivers - see temp_nrf5_mpsl - you can fetch ALL, but you cannot ask for ALL data. The same for bh1749, paw3212, sensor_sim.... Just all the drivers I had quickly look at.

File: samples/caf_sensor_manager/src/modules/sensor_stub_gen.c (hunk 4 above)
Comment (rakons) [reply samples/caf_sensor_manager/src/modules/sensor_stub_gen.c:170]: In the sample.yaml there is a check if the sensor_stub is working:
```
regex:
        - "sensor_stub_gen: READY"
        - "main state:READY"
 ```

File: samples/caf_sensor_manager/src/modules/sensor_stub_gen.c (hunk 6 above)
Comment (rakons) [reply samples/caf_sensor_manager/src/modules/sensor_stub_gen.c:15]: It is used for the correctness test.

File: samples/caf_sensor_manager/remote/CMakeLists.txt [hunk 12]
```diff
@@ -12,12 +12,11 @@ project(remote)
...
-  ../src/modules/sensor_sim_ctrl.c
-  ../configuration/common/event_proxy_init.c)
+target_sources(app PRIVATE src/main.c)
 
 # Add include directory for board specific CAF def files
 zephyr_include_directories(
   ../configuration/common
   ../configuration/${BOARD})
+
+  add_subdirectory(../src/modules "${CMAKE_CURRENT_BINARY_DIR}/modules")
```
[the remark this answers is not in the document]
Comment (rakons) [reply samples/caf_sensor_manager/remote/CMakeLists.txt]: Done

File: drivers/sensor/sensor_stub/Kconfig [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+# Copyright (c) 2018-2022 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+DT_COMPAT_NORDIC_SENSOR_STUB := nordic,sensor-stub
+
+config SENSOR_STUB
+	bool "Simplest sensor simulator"
+	default $(dt_compat_enabled,$(DT_COMPAT_NORDIC_SENSOR_STUB))
```
[the remark this answers is not in the document]
Comment (rakons) [reply drivers/sensor/sensor_stub/Kconfig:12]: Ok

File: drivers/sensor/sensor_stub/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,8 @@
+#
+# Copyright (c) 2018 Nordic Semiconductor
```
[the remark this answers is not in the document]
Comment (rakons) [reply drivers/sensor/sensor_stub/CMakeLists.txt:2]: Done

File: drivers/sensor/sensor_stub/Kconfig [hunk 15]
```diff
@@ -0,0 +1,14 @@
+# Sensor data simulator
+#
+# Copyright (c) 2018-2022 Nordic Semiconductor
```
[the remark this answers is not in the document]
Comment (rakons) [reply drivers/sensor/sensor_stub/Kconfig:3]: Done

File: drivers/sensor/sensor_stub/Kconfig (hunk 13 above)
[the remark this answers is not in the document]
Comment (rakons) [reply drivers/sensor/sensor_stub/Kconfig:12]: Done

File: samples/caf_sensor_manager/src/modules/Kconfig [hunk 16]
```diff
@@ -12,21 +12,38 @@ config SAMPLE_SENSOR_MANAGER_REMOTE_CORE_INITIALIZATION_TIMEOUT
...
-	default 10
-	help
-	  Simulated CPU busy time when processing sensor aggregator event.
-	  Sensor aggregator event comes with number of aggregated events.
-	  Simulated busy time is multiplied by this value.
+menuconfig WORKLOAD_SIM
+	bool "Simulated workload"
+if WORKLOAD_SIM
+
+	config WORKLOAD_TIME_PER_SAMPLE
```
[the remark this answers is not in the document]
Comment (rakons) [reply samples/caf_sensor_manager/src/modules/Kconfig:19]: I like that but I will update it to match the style of the rest config files.

File: samples/caf_sensor_manager/src/modules/sensor_stub_gen.c (hunk 7 above)
Comment (rakons) [reply samples/caf_sensor_manager/src/modules/sensor_stub_gen.c:86]: Well - this was exactly my idea - the function is Generate Channel, not Generate Changes. I do not think this is relative anyway. It could be even named gen_data... I do not have strong opinion here.

File: samples/caf_sensor_manager/remote/sensor_stub_overlay.conf (hunk 8 above)
Comment (rakons) [reply samples/caf_sensor_manager/remote/sensor_stub_overlay.conf:11]: We could, but then the build command would be more complex. Again - no strong opinion here.

File: samples/caf_sensor_manager/app.overlay (hunk 9 above)
Comment (rakons) [reply samples/caf_sensor_manager/app.overlay:12]: In my opinion they should. The fact that sensor_sim generates its driver even if there is not "okay" status seems like an error. It is possible only because sensor_sim is single instance driver. But I have already removed that from the sensor_sim as this is another change that is not needed in my implementation.

File: samples/caf_sensor_manager/prj.conf (hunk 10 above)
Comment (rakons) [reply samples/caf_sensor_manager/prj.conf:24]: I think it is fixed now.

File: samples/caf_sensor_manager/boards/nrf5340dk_nrf5340_cpuapp.conf (hunk 11 above)
Comment (rakons) [reply samples/caf_sensor_manager/boards/nrf5340dk_nrf5340_cpuapp.conf:21]: Right. This is a leftover from try of dts based configuration. Thank you for this catch.

File: samples/caf_sensor_manager/src/modules/sensor_stub_gen.c (hunk 7 above)
Comment (zycz) [reply samples/caf_sensor_manager/src/modules/sensor_stub_gen.c:86]: It's just a suggestion if you think it is ok leave it as it is.

File: samples/caf_sensor_manager/remote/sensor_stub_overlay.conf (hunk 8 above)
Comment (zycz) [reply samples/caf_sensor_manager/remote/sensor_stub_overlay.conf:11]: Again It's just a suggestion if you think it is ok leave it as it is. :)

**Review by zycz — APPROVED** [review]

File: samples/caf_sensor_manager/app.overlay (hunk 9 above)
Comment (zycz) [reply samples/caf_sensor_manager/app.overlay:12]: Discussed offline. Agreed this is currently best solution.

File: samples/caf_sensor_manager/README.rst [hunk 17]
```diff
@@ -32,6 +32,45 @@ This module handles the data received from the :ref:`caf_sensor_data_aggregator`
...
+=====================
+
+By default, the Sensor Manager sample uses Sensor Sim for sensor data simulation.
+You can select Sensor Stub module instead. This module has much simpler implementation,
+that relates fully on application implementing the function to generate the sensor data.
+This sample implements simple generator for Sensor Stub that uses no floating point mathematic.
+
+The Sensor Stub configuration is provided in the :file:`sensor_stub_overlay.conf` file.
+To use this configuration, specify the ``-DOVERLAY_CONFIG=sensor_stub_overlay.conf`` parameter along with the build command when building the sample.
+Note that for the multicore configuration it would change to ``-DremoteOVERLAY_CONFIG=sensor_stub_overlay.conf``:
```
[the remark this answers is not in the document]
Comment (rakons) [reply samples/caf_sensor_manager/README.rst]: Right. Done.

File: samples/caf_sensor_manager/src/modules/sensor_stub_gen.c (hunk 5 above)
Comment (emob-nordic) [reply samples/caf_sensor_manager/src/modules/sensor_stub_gen.c:110]: gosh... okay.

File: samples/caf_sensor_manager/src/modules/sensor_stub_gen.c (hunk 6 above)
Comment (emob-nordic) [reply samples/caf_sensor_manager/src/modules/sensor_stub_gen.c:15]: Understood

File: samples/caf_sensor_manager/README.rst [hunk 18]
```diff
@@ -32,6 +32,51 @@ This module handles the data received from the :ref:`caf_sensor_data_aggregator`
...
+
+Configuration
+*************
+
+|config|
+
+Single core configuration
+=========================
+
+For the MCU with multiple cores, the default configuration would use one core to simulate the sensor and the other core to process the sensor.
```
Comment (divipillai) [inline samples/caf_sensor_manager/README.rst:53]: @rakons : Is my understanding correct for these 2 sentences? I think we could have some explanation (one liner) about the single core configuration.

**Review by divipillai — COMMENTED** [review]
Also, shouldn't we have a change log entry?

File: samples/caf_sensor_manager/README.rst [hunk 19]
```diff
@@ -32,6 +32,51 @@ This module handles the data received from the :ref:`caf_sensor_data_aggregator`
...
+For the MCU that supports the single core configuration, the configurations are placed in the :file:`boards/<board>_singlecore.conf` file.
```
Comment (divipillai) [inline samples/caf_sensor_manager/README.rst:54]: @rakons : Is my understanding correct for these 2 sentences? I think we could have some explanation (one liner) about the single core configuration.

Comment (rakons) [reply samples/caf_sensor_manager/README.rst:54]:
```
For the MCU with multiple cores, the default configuration would use one core to simulate the sensor and the other core to process the sensor.
Such multiple core MCU may support single core configuration, where the sensor is simulated and processed on the single, selected core.
If this configuration is available it is placed in the :file:`boards/<board>_singlecore.conf` file.
```
?

Comment (divipillai) [reply samples/caf_sensor_manager/README.rst:54]: For the MCU with multiple cores, the default configuration would use one core to simulate the sensor and the other core to process the sensor.
These multiple core MCU can support single core configuration, where the sensor is simulated and processed on the single, selected core.
The configuration is placed in the :file:`boards/<board>_singlecore.conf` file.

**Review by emob-nordic — APPROVED** [review]

**Comment by rakons** [comment]
There was no tick from techwriter yet. The PR was not requested to merge. We have to fix documentation now.
